ddl/table.go

```diff
@@ -948,56 +956,188 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
+}
 
-	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+func checkPartitionReplica(partDefPointers []*model.PartitionDefinition, d *ddlCtx, job *model.Job) (needWait bool, err error) {
+	ctx := context.Background()
+	pdCli := d.store.(tikv.Storage).GetRegionCache().PDClient()
+	stores, err := pdCli.GetAllStores(ctx)
 	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+		return needWait, errors.Trace(err)
 	}
-	err = checkAddPartitionTooManyPartitions(uint64(len(tblInfo.Partition.Definitions) + len(partInfo.Definitions)))
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+	for _, pd := range partDefPointers {
+		startKey, endKey := tablecodec.GetTableHandleKeyRange(pd.ID)
+		regions, _, err := pdCli.ScanRegions(ctx, startKey, endKey, -1)
+		if err != nil {
+			return needWait, errors.Trace(err)
+		}
+		// For every region in the partition, if it has some corresponding peers and
+		// no pending peers, that means the replication has completed.
+		for _, region := range regions {
+			regionState, err := pdCli.GetRegionByID(ctx, region.Id)
+			if err != nil {
+				return needWait, errors.Trace(err)
+			}
+			tiflashPeerAtLeastOne := checkTiFlashPeerStoreAtLeastOne(stores, regionState.Meta.Peers)
+			// It's unnecessary to wait all tiflash peer to be replicated.
+			// Here only make sure that tiflash peer count > 0 (at least one).
+			if tiflashPeerAtLeastOne {
+				continue
+			}
+			needWait = true
+			logutil.BgLogger().Info("[ddl] partition replicas check failed in delete-only DDL state", zap.Int64("pID", pd.ID), zap.Uint64("wait region ID", region.Id), zap.Bool("tiflash peer at least one", tiflashPeerAtLeastOne), zap.Time("check time", time.Now()))
+			return needWait, nil
+		}
 	}
+	logutil.BgLogger().Info("[ddl] partition replicas check ok in delete-only DDL state")
+	return needWait, nil
+}
 
-	err = checkAddPartitionValue(tblInfo, partInfo)
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+func checkTiFlashPeerStoreAtLeastOne(stores []*metapb.Store, peers []*metapb.Peer) bool {
+	for _, peer := range peers {
+		for _, store := range stores {
+			if peer.StoreId == store.Id && storeHasEngineTiFlashLabel(store) {
+				return true
+			}
+		}
 	}
+	return false
+}
 
-	err = checkAddPartitionNameUnique(tblInfo, partInfo)
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+func storeHasEngineTiFlashLabel(store *metapb.Store) bool {
+	for _, label := range store.Labels {
+		if label.Key == "engine" && label.Value == "tiflash" {
+			return true
+		}
 	}
+	return false
+}
 
-	updatePartitionInfo(partInfo, tblInfo)
-	ver, err = updateVersionAndTableInfo(t, job, tblInfo, true)
+// TODO: It may have the issue when two clients concurrently add partitions to a table.
+func onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
```

Comment:
I think it should be moved to `partition.go`.

Reply:
it's same as what it was

Reply:
But it should be moved.

Reply:
addressed